src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+                // adds new timeouts in a loop.
+                for (int i = 0; i < 100000; i++)
+                {
+                    HashedWheelTimeout timeout;
+                    if (!this.owner.timeouts.TryDequeue(out timeout))
+                    {
+                        // all processed
+                        break;
+                    }
+                    if (timeout.State == HashedWheelTimeout.StCancelled)
+                    {
+                        // Was cancelled in the meantime.
+                        continue;
+                    }
+
+                    long calculated = (timeout.Deadline.Ticks + this.owner.tickDuration - 1) / this.owner.tickDuration; // ceiling to timeout later rather than earlier
+                    timeout.RemainingRounds = (calculated - this.tick) / this.owner.wheel.Length;
+
+                    long ticks = Math.Max(calculated, this.tick); // Ensure we don't schedule for past.
+                    int stopIndex = (int)(ticks & this.owner.mask);
+
+                    HashedWheelBucket bucket = this.owner.wheel[stopIndex];
+                    bucket.AddTimeout(timeout);
+                }
+            }
+
+            void ProcessCancelledTasks()
+            {
+                while (true)
+                {
+                    HashedWheelTimeout timeout;
+                    if (!this.owner.cancelledTimeouts.TryDequeue(out timeout))
+                    {
+                        // all processed
+                        break;
+                    }
+                    try
+                    {
+                        timeout.Remove();
+                    }
+                    catch (Exception ex)
+                    {
+                        if (Logger.WarnEnabled)
+                        {
+                            Logger.Warn("An exception was thrown while processing a cancellation task", ex);
+                        }
+                    }
+                }
+            }
+
+            /// <summary>
+            /// calculate timer firing time from startTime and current tick number,
+            /// then wait until that goal has been reached.
+            /// </summary>
+            /// <returns>long.MinValue if received a shutdown request,
+            /// current time otherwise (with long.MinValue changed by +1)
+            /// </returns>
+            TimeSpan WaitForNextTick()
+            {
+                long deadline = this.owner.tickDuration * (this.tick + 1);
+
+                while (true)
+                {
+                    TimeSpan currentTime = (PreciseTimeSpan.FromStart - this.owner.StartTime).ToTimeSpan();
+                    TimeSpan sleepTime = CeilTimeSpanToMilliseconds(TimeSpan.FromTicks(deadline - currentTime.Ticks));
+
+                    if (sleepTime <= TimeSpan.Zero)
+                    {
+                        if (currentTime.Ticks == long.MinValue)
```

Comment:
if this is happening consistently, then does that indicate that there are too many timers in the system to keep up with the tick count?